drivers/usb/bc12/bc12_pi3usb9201.c

```diff
@@ -0,0 +1,432 @@
+	}
+
+	/* Inform charge manager of new supplier type and current limit */
+	pi3usb9201_update_charging_partner(dev, &new_chg);
+}
+
+static int pi3usb9201_disconnect(const struct device *dev)
+{
+	int rv;
+
+	/* Ensure USB switch auto-on is enabled */
+	rv = pi3usb9201_bc12_usb_switch(dev, true);
+	if (rv < 0) {
+		return rv;
+	}
+
+	/* Put pi3usb9201 into its power down mode */
+	rv = pi3usb9201_set_mode(dev, PI3USB9201_POWER_DOWN);
+	if (rv < 0) {
+		return rv;
+	}
+
+	/* The start bc1.2 bit does not auto clear */
+	rv = pi3usb9201_bc12_detect_ctrl(dev, false);
+	if (rv < 0) {
+		return rv;
+	}
+
+	/* Mask interrupts until next bc1.2 detection event */
+	rv = pi3usb9201_interrupt_enable(dev, false);
+	if (rv < 0) {
+		return rv;
+	}
+
+	/*
+	 * Let the application know there's no more charge available for the
+	 * supplier type that was most recently detected.
+	 */
+	pi3usb9201_update_charging_partner(dev, NULL);
+
+	return 0;
+}
+
+static int pi3usb9201_set_portable_device(const struct device *dev)
+{
+	int rv;
+
+	/* Disable interrupts during mode change */
+	rv = pi3usb9201_interrupt_enable(dev, false);
+	if (rv < 0) {
+		return rv;
+	}
+
+	if (pi3usb9201_client_detect_start(dev) < 0) {
+		struct bc12_partner_state new_result;
+
+		/*
+		 * VBUS is present, but starting bc1.2 detection failed
+		 * for some reason. Set the partner type to unknown limit
+		 * current to the minimum allowed for a suspended USB device.
+		 */
+		new_result.voltage_uv = BC12_CHARGER_VOLTAGE_UV;
+		new_result.current_ua = BC12_CHARGER_MIN_CURR_UA;
+		new_result.type = BC12_TYPE_UNKNOWN;
+		/* Save supplier type and notify callbacks */
+		pi3usb9201_update_charging_partner(dev, &new_result);
+		LOG_ERR("bc1.2 detection failed, using defaults");
+
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void pi3usb9201_isr_work(struct k_work *item)
+{
+	struct pi3usb9201_data *pi3usb9201_data = CONTAINER_OF(item, struct pi3usb9201_data, work);
+	const struct device *dev = pi3usb9201_data->dev;
+	uint8_t client;
+	uint8_t host;
+	int rv;
+
+	rv = pi3usb9201_get_status(dev, &client, &host);
+	if (rv < 0) {
+		LOG_ERR("Failed to get host/client status");
+		return;
+	}
+
+	if (client != 0) {
+		/*
+		 * Any bit set in client status register indicates that
+		 * BC1.2 detection has completed.
+		 */
+		pi3usb9201_client_detect_finish(dev, client);
+	}
+}
+
+static void pi3usb9201_gpio_callback(const struct device *dev, struct gpio_callback *cb,
+				     uint32_t pins)
+{
+	struct pi3usb9201_data *pi3usb9201_data = CONTAINER_OF(cb, struct pi3usb9201_data, gpio_cb);
+
+	k_work_submit(&pi3usb9201_data->work);
+}
+
+static int pi3usb9201_set_role(const struct device *dev, const enum bc12_role role)
+{
+	switch (role) {
+	case BC12_DISCONNECTED:
+		return pi3usb9201_disconnect(dev);
+	case BC12_PORTABLE_DEVICE:
+		return pi3usb9201_set_portable_device(dev);
+
+	default:
+		LOG_ERR("unsupported BC12 role: %d", role);
+		return -EINVAL;
+	}
+
+	return 0;
```

Comment:
I would have thought that the static checkers would have caught this, unreachable statement